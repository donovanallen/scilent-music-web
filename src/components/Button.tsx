import { forwardRef } from 'react';

import { cn } from '@/lib/utils';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, children, disabled, type = 'button', ...props }, ref) => {
    return (
      <button
        type={type}
        disabled={disabled}
        className={cn(
          `w-fit rounded-md border border-transparent p-3 cursor-pointer disabled:cursor-not-allowed disabled:opacity-50 hover:opacity-75 transition text-dark bg-light subtitle`,
          className,
        )}
        ref={ref}
        {...props}
      >
        {children}
      </button>
    );
  },
);
Button.displayName = 'Button';

export default Button;
