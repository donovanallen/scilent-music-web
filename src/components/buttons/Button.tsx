import { LucideIcon } from 'lucide-react';
import * as React from 'react';
import { IconType } from 'react-icons';
import { ImSpinner2 } from 'react-icons/im';

import { cn } from '@/lib/utils';

const ButtonVariant = ['primary', 'outline', 'ghost', 'light', 'dark'] as const;
const ButtonSize = ['sm', 'base', 'lg'] as const;

type ButtonProps = {
  isLoading?: boolean;
  isDarkBg?: boolean;
  variant?: (typeof ButtonVariant)[number];
  size?: (typeof ButtonSize)[number];
  leftIcon?: IconType | LucideIcon;
  rightIcon?: IconType | LucideIcon;
  classNames?: {
    leftIcon?: string;
    rightIcon?: string;
  };
} & React.ComponentPropsWithRef<'button'>;

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      size = 'base',
      variant = 'primary',
      isDarkBg = false,
      leftIcon: LeftIcon,
      rightIcon: RightIcon,
      disabled: buttonDisabled,
      isLoading,
      children,
      className,
      classNames,
      ...rest
    },
    ref,
  ) => {
    const disabled = isLoading || buttonDisabled;

    return (
      <button
        ref={ref}
        type='button'
        disabled={disabled}
        className={cn(
          'inline-flex items-center rounded font-medium subtitle',
          'focus-visible:ring-primary-500 focus:outline-none focus-visible:ring',
          'shadow-sm',
          'transition duration-75',
          //#region  //*=========== Size ===========
          [
            size === 'lg' && ['px-6 py-4', 'text-lg'],
            size === 'base' && ['px-4 py-2', 'text-base'],
            size === 'sm' && ['px-2 py-1', 'text-xs md:text-sm'],
          ],
          //#endregion  //*======== Size ===========
          //#region  //*=========== Variants ===========
          [
            variant === 'primary' && [
              'bg-dark text-light hover:text-white active:text-white',
              'border-neutral-400 border',
              'disabled:opacity-80',
            ],
            variant === 'outline' && [
              'text-light hover:text-white active:text-white',
              'border-light border',
              'disabled:opacity-80',
              isDarkBg &&
                'bg-dark hover:bg-gray-900 active:bg-gray-800 disabled:bg-gray-800',
            ],
            variant === 'ghost' && [
              'text-light hover:text-white active:text-white',
              'shadow-none',
              'disabled:bg-opacity-80',
              isDarkBg &&
                'hover:bg-gray-900 active:bg-gray-800 disabled:bg-gray-800',
            ],
            variant === 'light' && [
              'bg-light text-dark hover:bg-white',
              'border border-gray-300',
              'active:bg-white/80 disabled:bg-gray-200',
            ],
            variant === 'dark' && [
              'bg-dark text-light hover:text-white active:text-white',
              'border-neutral-400 border',
              'disabled:opacity-80',
            ],
          ],
          //#endregion  //*======== Variants ===========
          'disabled:cursor-not-allowed',
          isLoading &&
            'relative text-transparent transition-none hover:text-transparent disabled:cursor-wait',
          className,
        )}
        {...rest}
      >
        {isLoading && (
          <div
            className={cn(
              'absolute left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2',
              {
                'text-white': ['primary', 'dark', 'ghost'].includes(variant),
                'text-black': ['light'].includes(variant),
                'text-primary-500': ['outline'].includes(variant),
              },
            )}
          >
            <ImSpinner2 className='animate-spin' />
          </div>
        )}
        {LeftIcon && (
          <div
            className={cn([
              size === 'lg' && 'mr-1.2',
              size === 'base' && 'mr-1',
              size === 'sm' && 'mr-1.5',
            ])}
          >
            <LeftIcon
              size='1em'
              className={cn(
                [
                  size === 'lg' && 'text-lg',
                  size === 'base' && 'text-base',
                  size === 'sm' && 'text-sm',
                ],
                classNames?.leftIcon,
              )}
            />
          </div>
        )}
        {children}
        {RightIcon && (
          <div
            className={cn([
              size === 'lg' && 'ml-1.2',
              size === 'base' && 'ml-1',
              size === 'sm' && 'ml-1.5',
            ])}
          >
            <RightIcon
              size='1em'
              className={cn(
                [
                  size === 'lg' && 'text-lg',
                  size === 'base' && 'text-base',
                  size === 'sm' && 'text-sm',
                ],
                classNames?.rightIcon,
              )}
            />
          </div>
        )}
      </button>
    );
  },
);

export default Button;
