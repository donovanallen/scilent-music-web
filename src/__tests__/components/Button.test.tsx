import { render } from '@testing-library/react';
import React from 'react';

import Button from '@/components/Button';

describe('Button', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });
});
