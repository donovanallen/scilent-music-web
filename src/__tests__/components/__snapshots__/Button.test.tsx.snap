// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<DocumentFragment>
  <button
    class="w-fit rounded-md border border-transparent p-3 cursor-pointer disabled:cursor-not-allowed disabled:opacity-50 hover:opacity-75 transition text-dark bg-light subtitle"
    type="button"
  />
</DocumentFragment>
`;
